### PR TITLE
Refactored and improved

### PR DESCRIPTION
**Pull request by xandao6**
I'm sorry, this PR got a little big, so take your time!

**The most important things:**
1. **seleniumUtils.py** - Designed for code reuse and make it easier to create new scripts. It will also help when we switch to pyppeteer.
1. **onlinetracker.py** - Many improvements to this script! Will help with issue #61. But I suggest wait for pyppeteer, asynchronous programming can help.

**Comments**
1. I couldn't test **tgbot.py** and **savechat.py** so I didn't make any big changes.
1. It's python3.6+ now, because I introduced the use of f-strings for string interpolation. 
Obs. Pyppetter is also 3.6+.
1. I only tested on Linux (Ubuntu 18.4).

### CODE REVIEW

File: wplay/seleniumUtils.py [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+from selenium.webdriver.common.keys import Keys
+from selenium.common.exceptions import NoSuchElementException,\
+    StaleElementReferenceException
+from webdriver_manager.chrome import ChromeDriverManager
+
+websites = {'whatsapp': 'https://web.whatsapp.com/'}
+
+
+def __get_XPATH_list(target=' '):
+    XPATH_list = {'wpp_target_title': f'//span[contains(@title, "{target}")]',
```
Comment (rpotter12) [inline wplay/seleniumUtils.py]: Codacy Issue found: [invalid syntax](https://app.codacy.com/manual/rpotter12/whatsapp-play/pullRequest?prid=4773095)

**Comment by rpotter12** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 1
- Added 9
           

Complexity increasing per file
==============================
- wplay/seleniumUtils.py  2
         

Clones removed
==============
+ wplay/onlinetracker.py  -1
+ wplay/messagetimer.py  -3
+ wplay/messageblast.py  -3
+ wplay/wchat.py  -1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/rpotter12/whatsapp-play/pullRequest?prid=4773095&bid=15904479)

File: wplay/onlinetracker.py [hunk 2]
```diff
@@ -1,77 +1,73 @@
...
 
-	# the name of the person by the user
-	target = str(name) #str(input("Enter the name of target: "))
+            if last_status != is_online:
+                # play sound when the person is online
+                if is_online == True:
+                    try:
+                        if is_sound_enabled == True:
+                            playsound('plucky.wav')
+                    except:
```
Comment (rpotter12) [inline wplay/onlinetracker.py]: Codacy Issue found: [No exception type(s) specified](https://app.codacy.com/manual/rpotter12/whatsapp-play/pullRequest?prid=4773095)

File: wplay/onlinetracker.py [hunk 3]
```diff
@@ -1,77 +1,73 @@
...
-from webdriver_manager.chrome import ChromeDriverManager
-
-# to play sound
-from playsound import playsound
-
-# for system control
 import time
 import os
-import datetime
+import subprocess
```
Comment (rpotter12) [inline wplay/onlinetracker.py]: Codacy Issue found: [Consider possible security implications associated with subprocess module.](https://app.codacy.com/manual/rpotter12/whatsapp-play/pullRequest?prid=4773095)

File: wplay/savechat.py [hunk 4]
```diff
@@ -100,28 +124,37 @@ def localFileList():
...
         if entries['f'] == asset:
             return entries['f'], entries['m'], entries['r'], entries['s']
 
+
 def getMultipleFiles(data, folder):
     files = localFileList()
     data = json.loads(data)
     for entries in data:
-        if any(entries['m'] in lists for lists in files) == False or 'database' in entries['f'].lower():
+        if any(entries['m'] in lists for lists in files) == \
```
Comment (rpotter12) [inline wplay/savechat.py]: Codacy Issue found: [comparison to False should be 'if cond is False:' or 'if not cond:'](https://app.codacy.com/manual/rpotter12/whatsapp-play/pullRequest?prid=4773095)

File: wplay/messagetimer.py [hunk 5]
```diff
@@ -28,35 +22,22 @@ def msgTimer(name):
...
-    i = 0
-    while i < n:
+    for _ in range(number_of_messages):
         if not messages:
             break
         else:
-            message_area.send_keys(
-                messages[random.randrange(0, nMessages)] + Keys.ENTER)
+            sel.send_message(
+                message_area, messages[random.randrange(0, message_type_numbers)])
```
Comment (rpotter12) [inline wplay/messagetimer.py]: Codacy Issue found: [Standard pseudo-random generators are not suitable for security/cryptographic purposes.](https://app.codacy.com/manual/rpotter12/whatsapp-play/pullRequest?prid=4773095)

File: wplay/onlinetracker.py [hunk 6]
```diff
@@ -1,77 +1,73 @@
...
+                    status = 'offline'
+            except:
+                status = 'offline'
+                is_online = False
 
-	# the name of the person by the user
-	target = str(name) #str(input("Enter the name of target: "))
+            if last_status != is_online:
+                # play sound when the person is online
+                if is_online == True:
```
Comment (rpotter12) [inline wplay/onlinetracker.py]: Codacy Issue found: [comparison to True should be 'if cond is True:' or 'if cond:'](https://app.codacy.com/manual/rpotter12/whatsapp-play/pullRequest?prid=4773095)

File: wplay/onlinetracker.py [hunk 7]
```diff
@@ -1,77 +1,73 @@
...
+                    try:
+                        if is_sound_enabled == True:
```
Comment (rpotter12) [inline wplay/onlinetracker.py]: Codacy Issue found: [comparison to True should be 'if cond is True:' or 'if cond:'](https://app.codacy.com/manual/rpotter12/whatsapp-play/pullRequest?prid=4773095)

File: wplay/onlinetracker.py [hunk 8]
```diff
@@ -1,77 +1,73 @@
...
+
+    sel.find_and_navigate_to_target(driver_wait, chosen_website, name)
+
+    # finds if online_status directory is present
+    if 'tracking_data' not in os.listdir(os.getcwd()):
+        os.mkdir('tracking_data')
+
+    # create status.txt file and overwrite if exists
+    status_file = open(
+        os.path.join('tracking_data', f'status_{name}.txt'), 'w')
```
Comment (rpotter12) [inline wplay/onlinetracker.py]: Codacy Issue found: [invalid syntax](https://app.codacy.com/manual/rpotter12/whatsapp-play/pullRequest?prid=4773095)

File: wplay/onlinetracker.py (hunk 3 above)
Comment (rpotter12) [inline wplay/onlinetracker.py]: Codacy Issue found: [Unused import subprocess](https://app.codacy.com/manual/rpotter12/whatsapp-play/pullRequest?prid=4773095)

**Comment by rpotter12** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 1
- Added 5
           

Complexity increasing per file
==============================
- wplay/seleniumUtils.py  2
         

Clones removed
==============
+ wplay/onlinetracker.py  -1
+ wplay/wchat.py  -1
+ wplay/messagetimer.py  -3
+ wplay/messageblast.py  -3
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/rpotter12/whatsapp-play/pullRequest?prid=4773095&bid=15904479)

File: wplay/seleniumUtils.py [hunk 9]
```diff
@@ -0,0 +1,59 @@
...
+from selenium.common.exceptions import NoSuchElementException,\
+    StaleElementReferenceException
+from webdriver_manager.chrome import ChromeDriverManager
+
+websites = {'whatsapp': 'https://web.whatsapp.com/'}
+
+
+def __get_XPATH_list(target=' '):
+    XPATH_list = {
+        'wpp_target_title': f'//span[contains(@title, "{target}")]',
```
Comment (rpotter12) [inline wplay/seleniumUtils.py]: Codacy Issue found: [invalid syntax](https://app.codacy.com/manual/rpotter12/whatsapp-play/pullRequest?prid=4773095)

File: wplay/onlinetracker.py [hunk 10]
```diff
@@ -1,77 +1,74 @@
...
+
+    sel.find_and_navigate_to_target(driver_wait, chosen_website, name)
+
+    # finds if online_status directory is present
+    if 'tracking_data' not in os.listdir(os.getcwd()):
+        os.mkdir('tracking_data')
+
+    # create status.txt file and overwrite if exists
+    status_file = open(
+        os.path.join('tracking_data', f'status_{name}.txt'), 'w'
```
Comment (rpotter12) [inline wplay/onlinetracker.py]: Codacy Issue found: [invalid syntax](https://app.codacy.com/manual/rpotter12/whatsapp-play/pullRequest?prid=4773095)

File: wplay/messagetimer.py (hunk 5 above)
Comment (rpotter12) [inline wplay/messagetimer.py]: Codacy Issue found: [Standard pseudo-random generators are not suitable for security/cryptographic purposes.](https://app.codacy.com/manual/rpotter12/whatsapp-play/pullRequest?prid=4773095)

File: wplay/onlinetracker.py [hunk 11]
```diff
@@ -1,77 +1,74 @@
...
 
-	# the name of the person by the user
-	target = str(name) #str(input("Enter the name of target: "))
+            if last_status != is_online:
+                # play sound when the person is online
+                if is_online:
+                    try:
+                        if is_sound_enabled:
+                            playsound('plucky.wav')
+                    except:
```
Comment (rpotter12) [inline wplay/onlinetracker.py]: Codacy Issue found: [No exception type(s) specified](https://app.codacy.com/manual/rpotter12/whatsapp-play/pullRequest?prid=4773095)

File: wplay/savechat.py (hunk 4 above)
Comment (rpotter12) [inline wplay/savechat.py]: Codacy Issue found: [comparison to False should be 'if cond is False:' or 'if not cond:'](https://app.codacy.com/manual/rpotter12/whatsapp-play/pullRequest?prid=4773095)

File: wplay/seleniumUtils.py (hunk 9 above)
Comment (xandao6) [reply wplay/seleniumUtils.py]: I don't see anything wrong here.

File: wplay/onlinetracker.py (hunk 10 above)
Comment (xandao6) [reply wplay/onlinetracker.py]: I don't see anything wrong here.

File: wplay/savechat.py (hunk 4 above)
Comment (xandao6) [reply wplay/savechat.py]: I don't know if we change to 'if not cond' will break or not, because I wasn't able to test this script.

**Pull request merged**